schema-engine/commands/src/commands/diff.rs

```diff
@@ -0,0 +1,192 @@
+use crate::{
+    CoreError, SchemaContainerExt,
+    core_error::CoreResult,
+    dialect_for_provider, extract_namespaces,
+    json_rpc::types::{DiffParams, DiffResult, DiffTarget, UrlContainer},
+};
+use enumflags2::BitFlags;
+use json_rpc::types::MigrationList;
+use psl::SourceFile;
+use schema_connector::{
+    ConnectorError, DatabaseSchema, ExternalShadowDatabase, Namespaces, SchemaConnector, SchemaDialect,
+};
+
+pub async fn diff(params: DiffParams, connector: &mut dyn SchemaConnector) -> CoreResult<DiffResult> {
+    // In order to properly handle MultiSchema, we need to make sure the preview feature is
+    // correctly set, and we need to grab the namespaces from the Schema, if any.
+    // Note that currently, we union all namespaces and preview features. This may not be correct.
+    // TODO: This effectively reads and parses (parts of) the schema twice: once here, and once
+    // below, when defining 'from'/'to'. We should revisit this.
+    let (namespaces, preview_features) =
+        namespaces_and_preview_features_from_diff_targets(&[&params.from, &params.to])?;
+
+    let from = diff_target_to_dialect(
+        &params.from,
+        params.shadow_database_url.as_deref(),
+        connector,
+        namespaces.clone(),
+        preview_features,
+    )
+    .await?;
+    let to = diff_target_to_dialect(
+        &params.to,
+        params.shadow_database_url.as_deref(),
+        connector,
+        namespaces,
+        preview_features,
+    )
+    .await?;
+
+    // The `from` connector takes precedence, because if we think of diffs as migrations, `from` is
+    // the target where the migration would be applied.
+    //
+    // TODO: make sure the connectors are the same in from and to.
+    let (dialect, from, to) = match (from, to) {
+        (Some((connector, from)), Some((_, to))) => (connector, from, to),
+        (Some((connector, from)), None) => {
+            let to = connector.empty_database_schema();
+            (connector, from, to)
+        }
+        (None, Some((connector, to))) => {
+            let from = connector.empty_database_schema();
+            (connector, from, to)
+        }
+        (None, None) => {
+            return Err(ConnectorError::from_msg(
+                "Could not determine the connector to use for diffing.".to_owned(),
+            ));
+        }
+    };
```

Comment:
might be a more concise way of writing this
```suggestion
        let (conn_from, schema_from) = diff_target_to_dialect(...).await?.unzip();
        let (conn_to, schema_to) = diff_target_to_dialect(...).await?.unzip();

        let Some(conn) = conn_from.or(conn_to) {
            return Err(ConnectorError::from_msg(
                "Could not determine the connector to use for diffing.".to_owned(),
            ));
        }
        let from = b.unwrap_or_else(|| conn.empty_database_schema());
        let to = d.unwrap_or_else(|| conn.empty_database_schema());
```
```

Reply:
This was also copied verbatim from a previous `diff` implementation btw

Reply:
`b` and `d` are non-existent symbols

Reply:
Did you mean something like:

```rust

    let (conn_from, schema_from) = diff_target_to_dialect(
        &params.from,
        params.shadow_database_url.as_deref(),
        connector,
        namespaces.clone(),
        preview_features,
    )
    .await?
    .unzip();

    let (conn_to, schema_to) = diff_target_to_dialect(
        &params.to,
        params.shadow_database_url.as_deref(),
        connector,
        namespaces,
        preview_features,
    )
    .await?
    .unzip();

    let dialect = conn_from
        .or(conn_to)
        .ok_or_else(|| ConnectorError::from_msg("Could not determine the connector to use for diffing.".to_owned()))?;

    let from = schema_from.unwrap_or_else(|| dialect.empty_database_schema());
    let to = schema_to.unwrap_or_else(|| dialect.empty_database_schema());
```

?

Reply:
yeah that's right, sorry, quickly written it down in the github comment